### PR TITLE
WIP: Introduce an enhancement proposal process

### PR DESCRIPTION
**Pull request by ironcladlou**
This commit introduces an enhancement proposal process copied from
the OpenShift enhancements process but stripped down to keep only
the pieces which are currently relevant to HyperShift.

### CODE REVIEW

**Comment by ironcladlou** [comment]
So far I have only stripped away stuff from the OCP content which I thought was inapplicable, and haven't added anything HyperShift specific. Suggestions welcome...

File: enhancements/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+## Life-cycle
+
+
+Pull requests to this repository should be short-lived and merged as
+soon as there is consensus.
+
+Ideally pull requests with enhancement proposals will be merged before
+significant coding work begins, since this avoids having to rework the
+implementation if the design changes as well as arguing in favor of
+accepting a design simply because it is already implemented.
```
Comment (alvaroaleman) [inline enhancements/README.md:78]: Not sure if that is maybe too detailed already, but should we add a "constraints to consider" or such section that mentions for example the fact that we can not use webhooks for performance reasons?

File: enhancements/guidelines/enhancement_template.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+- If an `N->N+1` upgrade fails mid-way through, or if the `N+1` cluster is
+  misbehaving, it should be possible for the user to rollback to `N`. It is
+  acceptable to require some documented manual steps in order to fully restore
+  the downgraded cluster to its previous state. Examples of acceptable steps
+  include:
+  - Deleting any CVO-managed resources added by the new version. The
+    CVO does not currently delete resources that no longer exist in
+    the target version.
+
+### Version Skew Strategy
```
Comment (alvaroaleman) [inline enhancements/guidelines/enhancement_template.md:186]: Maybe mention "Hypershift operator might end up running a different version of the controlplane operator based on cluster version"?

Comment (ironcladlou) [reply enhancements/guidelines/enhancement_template.md:186]: Good idea, I removed some of the points here and replaced them with a blurb about this aspect of HyperShift. What do you think?

File: enhancements/README.md (hunk 1 above)
Comment (ironcladlou) [reply enhancements/README.md:78]: I like that idea, although I'm also unsure where to put it. Any other "global constraints" we already know if? Maybe if there are more than one they could justify a section somewhere to make them more obvious. cc @csrwng @enxebre @sjenning @derekwaynecarr

File: enhancements/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+- update status of the enhancement effort
+- link to relevant issues in other repos
+
+Please do not comment on the enhancement issue to:
+- discuss a detail of the design, code or docs. Use a linked-to-issue or design PR
+  for that
+
+## Life-cycle
+
+
```
Comment (awgreene) [inline enhancements/README.md:71]: Nit: Please remove the extra newline.

File: enhancements/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+## When to Comment on an Enhancement Issue
+
+Please comment on the enhancement issue to:
+- request a review or clarification on the process
+- update status of the enhancement effort
+- link to relevant issues in other repos
+
+Please do not comment on the enhancement issue to:
+- discuss a detail of the design, code or docs. Use a linked-to-issue or design PR
```
Comment (awgreene) [inline enhancements/README.md:66]: This is interesting - is it standard process not to comment on design in an enhancement? There is a proposed section in the template called [Implementation Details/Notes/Constraints](https://github.com/openshift/hypershift/pull/511/files#diff-a4eb7b31979a31e26a883f2b3e7747a0f804969d5c780ed1002ec99141f6a595R105). If I have comments on the proposed design, am I supposed to create a new issue and point it to the enhancement PR and hope that it is merged after my issue is addressed?

**Review by awgreene — APPROVED** [review]
Nice work @ironcladlou - non-blocking comments.

File: enhancements/README.md (hunk 4 above)
Comment (ironcladlou) [reply enhancements/README.md:66]: I was confused by this too, but upon further inspection, the blurb is referring to the enhancement _issue_ which is something different from the enhancement _PR_. The implication seems to be that the enhancement starts as an _issue_, leading to a _PR from the template_, and the _issue_ is apparently for tracking progress at the high level.

However, the readme doesn't describe the process of creating an issue, just a PR from the template. So maybe we should purge all references to the issue? Or if we want the issue -> PR process, maybe we should clarify?

Comment (awgreene) [reply enhancements/README.md:66]: Ah - I see the distinction now. Requiring an issue to be created first will probably help non-contributors request a feature, but it also introduces additional overhead if its a hard requirement. I lean towards making projects more inclusive from the onset, but will let other chime in.

Comment (ironcladlou) [reply enhancements/README.md:66]: I agree with you that the less overhead the better, we can always introduce an issue to the process later if we decide that would make any sense.

Comment (ironcladlou) [reply enhancements/README.md]: I went ahead and removed this section entirely

File: enhancements/guidelines/enhancement_template.md (hunk 2 above)
Comment (alvaroaleman) [reply enhancements/guidelines/enhancement_template.md:186]: looks good, thanks

File: enhancements/README.md (hunk 1 above)
Comment (ironcladlou) [reply enhancements/README.md:66]: @csrwng great list, thank you. I incorporated those (and @alvaroaleman's webhook point) into a "General constraints" section, PTAL.

File: enhancements/guidelines/enhancement_template.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+- What additional testing is necessary to support managed OpenShift service-based offerings?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation and anything particularly
+challenging to test should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations).
+
+### Upgrade / Downgrade Strategy
```
[the remark this answers is not in the document]
Comment (csrwng) [reply enhancements/guidelines/enhancement_template.md:146]: @ironcladlou the one thing that occurs to me is that for any code added to hypershift we have to decide what its version lifecycle is going to be. Unlike a standalone OCP cluster we have components with different lifecycles:
- the hypershift operator
- CAPI and related resources
- OCP release along with control plane operator

File: enhancements/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+implementation if the design changes as well as arguing in favor of
+accepting a design simply because it is already implemented.
+
+## General constraints
+
+HyperShift has some general constraints which are outlined here in order to help
+inform proposals. Every proposal should take these constraints into
+consideration.
+
+1. Webhooks are generally avoided for performance and reliability
```
Comment (csrwng) [inline enhancements/README.md:74]: It may be good to specify that this is webhooks on the management cluster. Webhooks on the guest clusters should be ok.

Comment (ironcladlou) [reply enhancements/README.md]: Fixed

**Pull request closed without merging**